ethr/ethr.go

```diff
@@ -0,0 +1,473 @@
+	}
+	networkID := "dev"
+	var chainID *big.Int
+	id := fullID[2]
+	if fullID[1] != "" {
+		if !strings.HasSuffix(fullID[1], ":") {
+			return nil, fmt.Errorf("not a valid ethr did: %s", did)
+		}
+		networkID = fullID[1][:len(fullID[1])-1]
+	}
+	var ok bool
+	// First check to see if this networkID matches a known plaintext network name
+	chainID, ok = knownNetworks[networkID]
+	if !ok {
+		// If it doesn't, we assume it is a hex chainID code and work with that
+		chainID = new(big.Int)
+		chainID.SetString(networkID, 16)
+	}
+	if chainID.Cmp(r.chainID) != 0 {
+		// TODO: Rather than error, we'll want to support multiple chains
+		return nil, fmt.Errorf("non-matching chain ids")
+	}
+
+	controller, history, publicKey, err := changeLog(r.client, id)
+	if err != nil {
+		return nil, err
+	}
+	return wrapDocument(did, controller, publicKey, history, r.chainID)
+}
+
+var _ resolver.Resolver = (*Resolver)(nil)
+
+// EthereumDIDRegistryDIDEventUnion represents any of the available DIDRegistry
+// event types: *OwnerChanged, *DelegateChanged, *AttributeChange
+type EthereumDIDRegistryDIDEventUnion struct {
+	Identity       common.Address
+	Owner          common.Address
+	DelegateType   [32]byte
+	Delegate       common.Address
+	ValidTo        *big.Int
+	PreviousChange *big.Int
+	Raw            types.Log // Blockchain specific contextual infos
+	Name           [32]byte
+	Value          []byte
+	Type           string
+}
+
+func wrapDocument(did string, controller *common.Address, publicKey *ecdsa.PublicKey, history []*EthereumDIDRegistryDIDEventUnion, chainID *big.Int) (*resolver.Document, error) {
+	now := big.NewInt(time.Now().Unix())
+	verificationMethod := []resolver.VerificationMethod{{
+		ID:                  fmt.Sprintf("%s#controller", did),
+		Type:                "EcdsaSecp256k1RecoveryMethod2020",
+		Controller:          did,
+		BlockchainAccountID: fmt.Sprintf("%s@eip155:%d", controller.Hex(), chainID),
+	}}
+
+	authentication := []string{fmt.Sprintf("%s#controller", did)}
+
+	if publicKey != nil {
+		// NOTE: This is the uncompressed key bytes
+		publicKeyBytes := crypto.FromECDSAPub(publicKey)
+		publicKeyMultibase, err := multibase.Encode(multibase.Base16, publicKeyBytes)
+		if err != nil {
+			return nil, err
+		}
+		verificationMethod = append(verificationMethod, resolver.VerificationMethod{
+			ID:                 fmt.Sprintf("%s#controllerKey", did),
+			Type:               "EcdsaSecp256k1VerificationKey2019",
+			Controller:         did,
+			PublicKeyMultibase: publicKeyMultibase,
+		})
+
+		authentication = append(authentication, fmt.Sprintf("%s#controllerKey", did))
+	}
+
+	deactivated := false
+	auths := orderedmap.New()
+	pks := orderedmap.New()
+	services := orderedmap.New()
+	delegateCount := 0
+	serviceCount := 0
+
+	for _, event := range history {
+		validTo := event.ValidTo
+		// fmt.Println(validTo.Uint64(), now.Uint64())
+		if (validTo != nil) && validTo.Cmp(now) >= 0 {
+			switch event.Type {
```

Comment:
I see what you were mentioning now in slack. The events come from the chain right? In this case, I think it would be cleaner to hard type each event and encode history as `[]interface{}`. Then here you can switch on `event.(type)`. That would mean anyone reading this code knows which event params below to each type. I.e, you would have an event like,

```
type OwnerChanged struct {
    ...
}
```